src/navigation/failure_identification/watchdog.py

```diff
@@ -1,11 +1,97 @@
 import numpy as np
 import rospy
 from pandas import DataFrame
+from util.ros_utils import get_rosparam
+from util.SO3 import SO3
+from typing import Tuple
+
+WINDOW_SIZE = get_rosparam("watchdog/window_size", 100)
+ANGULAR_THRESHOLD = get_rosparam("watchdog/angular_threshold", 0.001)
+LINEAR_THRESHOLD = get_rosparam("watchdog/linear_threshold", 0.55)
 
 
 class WatchDog:
-    def __init__(self):
-        pass
+    def __init__(self, collector_in):
+        self.collector = collector_in
+
+    def get_start_end_positions(self, dataframe: DataFrame) -> Tuple[np.ndarray, np.ndarray]:
+        # get the average of the first 25% of the dataframes x, y, z position from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_x = np.mean(start["x"])
+        start_y = np.mean(start["y"])
+        start_z = np.mean(start["z"])
+        start_pos = np.array([start_x, start_y, start_z])
+        end_x = np.mean(end["x"])
+        end_y = np.mean(end["y"])
+        end_z = np.mean(end["z"])
+        end_pos = np.array([end_x, end_y, end_z])
+        return start_pos, end_pos
+
+    def get_start_end_rotations(self, dataframe: DataFrame) -> Tuple[SO3, SO3]:
+        # get the average of the first 25% of the dataframes rotation from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_rot = np.array(
+            [np.mean(start["rot_x"]), np.mean(start["rot_y"]), np.mean(start["rot_z"]), np.mean(start["rot_w"])]
```

Comment:
unfortunately, I'm pretty sure you can't just average all the values of a quaternion to get the average rotation. Looks like there is a more complex algorithm for averaging quaternions if you Google it

Reply:
Let say you have a set of rotations S. I believe the correct way is to take the vee(log(R)) for each R in S and compute the arithmetic mean of those vecs. Call the resulting vector v. Then take exp(hat(v)) to get the average rotation.   

manif or Sophus would've saved us here. Instead this is going to be painful.